File: LAB QUIZZES/CSE221_Quiz04/_A4D/_A4D.py
import os, sys, time, random, threading, subprocess
from _A4D_Solution import solve
start = 0; finish = 0; elapsed = 0
test = 0; best = 0; score = 0; batch = 0; total = 0
timeLimit = 1; nBatch = 13; N = 0; 
weight = [ 0, 6, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8]
nTest =  [0, 1, 1, 1, 10000, 10000, 1000, 1000, 100, 100, 50, 50, 20, 20]
maxN =  [ 0, 5, 5, 8, 10,  10, 100, 100, 1000, 1000, 10000, 10000, 100000, 100000]
res = False
perm = []
edges = []
o = 0

def printCase(verdict, test, N, edges):
    if os.path.getsize(verdict + ".txt") == 0:
        with open(verdict + ".txt", "w") as fout:
            fout.write("batch = " + str(batch) + "\ntest = " + str(test) + "\n")
            fout.write("N = " + str(N) + "\n")
            fout.write("edges = " + str(edges) + "\n")
    print(verdict + " on Batch " + str(batch), flush=True)
def exitCase(verdict):
    printCase(verdict, test, N, edges)
    os._exit(1)
def checkSoln():
    global start, finish, elapsed, N, edges, o
    start = time.time()
    res = solve(N, edges)
    finish = time.time()
    elapsed += finish - start
    o = (o == 0 or o == 2)
    if o != res:
        exitCase("WrongAnswer")
def processBatch():
    global test, N, edges, o, perm
   
    if batch == 1:
        N = 5
        edges = [(0, 1), (0, 2), (0, 3), (0, 4), (1, 2), (1, 3), (2, 3), (2, 4), (3, 4)]
        o = 0
        checkSoln()
        return
    elif batch == 2:
        N = 5
        edges = [(0, 1), (0, 2), (0, 3), (2, 3), (2, 4), (3, 4)]
        o = 1
        checkSoln()
        return
    elif batch == 3:
        N = 8
        edges = [(0, 2), (0, 4), (1, 0), (1, 4), (1, 6), (1, 5), (2, 4), (2, 5), (3, 2), (4, 3), (4, 7), (5, 3), (6, 4), (6, 2), (6, 5), (7, 0)]
        o = 0
        checkSoln()
        return
    test = 1
    while test <= nTest[batch]:
        N = random.randint(1, maxN[batch])
        M = random.randint(N - 1, 2 * maxN[batch])
        perm = [0] * N
        par = [0] * N
        edges = []
        o = 0
        for i in range(N):
            perm[i] = i
        random.shuffle(perm)
        for i in range(1, N):
            u = perm[random.randint(0, i - 1)]
            v = perm[i]
            par[u]^=1
            par[v]^=1
            if(par[u] == par[v]):
                o += 4 * par[u] - 2
            edges.append((u, v))
            M -= 1
        A = random.randint(0, 2)
        if(N <= 3 or A == 2):
            while(M > N / 2):
                u = random.randint(0, N - 1)
                v = random.randint(0, N - 1)
                par[u]^=1
                par[v]^=1
                if(u != v and par[u] == par[v]):
                    o += 4 * par[u] - 2
                edges.append((u, v))
                M -= 1
            oo = []
            for i in range(N):
                if(par[i] == 1):
                    oo.append(i)
            random.shuffle(oo)
            for i in range(0, len(oo), 2):
                u = oo[i]
                v = oo[i + 1]
                par[u]^=1
                par[v]^=1
                if(u != v and par[u] == par[v]):
                    o += 4 * par[u] - 2
                edges.append((u, v))
                M -= 1
        else:
            while(M > A):
                u = random.randint(0, N - 1)
                v = random.randint(0, N - 1)
                par[u]^=1
                par[v]^=1
                if(u != v and par[u] == par[v]):
                    o += 4 * par[u] - 2
                edges.append((u, v))
                M -= 1
            if(A == 1 and o <= 2):
                ee = []
                for i in range(N):
                    if(par[i] == 0):
                        ee.append(i)
                u = ee[0]
                v = ee[1]
                par[u]^=1
                par[v]^=1
                if(u != v and par[u] == par[v]):
                    o += 4 * par[u] - 2
                edges.append((u, v))
                M -= 1
        checkSoln()
        test += 1

def limitTime():
    recall = -1
    while elapsed < timeLimit and recall != elapsed:
        recall = elapsed
        time.sleep(timeLimit << 1)
    exitCase("TimeLimitExceeded")
if len(sys.argv) == 2:
    batch = int(sys.argv[1])
    random.seed(batch)
if 1 <= batch and batch <= nBatch:
    print("Running on Batch " + str(batch), flush=True)
    threading.Thread(target=limitTime, daemon=True).start()
    processBatch()
    print("Accepted on Batch " + str(batch) + " in " + f"{elapsed:.9f}" + "s", flush=True)
    os._exit(0)
ID = ""
with open("EnterIDandLanguage.txt") as fin:
    ID = fin.readline().split()[0]
if os.path.exists("_A4D_" + ID + ".py"):
    with open("_A4D_" + ID + ".py") as fin:
        best = int(fin.readline().split()[2])
open("WrongAnswer.txt", "w").close()
open("TimeLimitExceeded.txt", "w").close()
batch = 1
while batch <= nBatch:
    process = subprocess.run("pypy _A4D.py " + str(batch), capture_output=True, text=True)
    if process.stdout:
        print(process.stdout, end="")
    if process.stderr:
        print(process.stderr, end="")
    if process.returncode == 0:
        score += weight[batch]
    total += weight[batch]
    batch += 1
if best <= score:
    with open("_A4D_" + ID + ".py", "w") as fout:
        fout.write("# " + ID + " " + str(score) + "\n")
    subprocess.run("cmd /c echo # %COMPUTERNAME% %USERNAME%>>_A4D_" + ID + ".py")
    with open("_A4D_" + ID + ".py", "a") as fout:
        with open("_A4D_Solution.py") as code:
            fout.write(code.read())
print("Tentative score = " + str(score / total) + "/1", flush=True)
os._exit(0)

File: LAB QUIZZES/CSE221_Quiz04/_A4D/_A4D.java
import java.util.*;
import java.io.*;

public class _A4D {
    public static Random rng = new Random();
    public static long start, finish, elapsed = 0, MOD = 1000003;
    public static int test = 0, best = 0, score = 0, batch = 0, total = 0;
    public static int timeLimit = 1, nBatch = 13;
    public static int weight[] = {  0, 6, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8};
    public static int nTest[] =  {0, 1, 1, 1, 10000, 10000, 1000, 1000, 100, 100, 50, 50, 20, 20};
    public static int maxN[] =  {0, 5, 5, 8, 10,  10, 100, 100, 1000, 1000, 10000, 10000, 100000, 100000};
    public static int N, o;
    public static boolean res;
    public static ArrayList<_A4D_Solution.Pair<Integer, Integer>> edges;
    public static ArrayList<Integer> par, perm;

    public static void printCase(String verdict, int test, int N, ArrayList<_A4D_Solution.Pair<Integer, Integer>> edges)
            throws Exception {
        if (new File(verdict + ".txt").length() == 0) {
            BufferedWriter fout = new BufferedWriter(new FileWriter(verdict + ".txt"));
            fout.write("batch = " + batch + "\ntest = " + test + "\n");
            fout.write("N = " + N + "\n");
            fout.write("edges = " + edges + "\n");
            fout.close();
        }
        System.out.print(verdict + " on Batch " + batch + "\n");
    }

    public static void exitCase(String verdict) throws Exception {
        printCase(verdict, test, N, edges);
        System.exit(1);
    }

    public static void checkSoln() throws Exception {
        start = System.nanoTime();
        res = _A4D_Solution.solve(N, edges);
        finish = System.nanoTime();
        elapsed += finish - start;
        if (1000000000L * timeLimit < elapsed)
            exitCase("TimeLimitExceeded");
        boolean a = (o == 0 || o == 2);
        if(res != a)
            exitCase("WrongAnswer");
    }

    public static void processBatch() throws Exception {
        if(batch == 1){
            N = 5;
            edges = new ArrayList<>();
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 1));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 2));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(1, 2));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(1, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(2, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(2, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(3, 4));
            o = 0;
            checkSoln();
            return; 
        }else if(batch == 2){
            N = 5;
            edges = new ArrayList<>();
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 1));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 2));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(2, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(2, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(3, 4));
            o = 1;
            checkSoln();
            return;
        }else if(batch == 3){
            N = 8;
            edges = new ArrayList<>();
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 2));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(0, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(1, 0));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(1, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(1, 6));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(1, 5));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(2, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(2, 5));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(3, 2));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(4, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(4, 7));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(5, 3));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(6, 4));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(6, 2));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(6, 5));
            edges.add(new _A4D_Solution.Pair<Integer, Integer>(7, 0));
            o = 0;
            checkSoln();
            return;
        }
        for (test = 1; test <= nTest[batch]; ++test) {
            N = rng.nextInt(maxN[batch]) + 1;
            int M = rng.nextInt(2 * maxN[batch] - N + 1) + N - 1;
            perm = new ArrayList<Integer>();
            par = new ArrayList<Integer>();
            edges = new ArrayList<_A4D_Solution.Pair<Integer, Integer>>();
            o = 0;
            for(int i = 0; i < N; i++){
                perm.add(i);
                par.add(0);
            }
            Collections.shuffle(perm);
            for(int i = 1; i < N; i++){
                int u = perm.get(rng.nextInt(i));
                int v = perm.get(i);
                par.set(u, par.get(u) ^ 1); par.set(v, par.get(v) ^ 1);
                if(par.get(u) == par.get(v))o += 4 * par.get(u) - 2;
                edges.add(new _A4D_Solution.Pair<Integer, Integer>(u, v));
                M--;
            }
            int A = rng.nextInt(3);
            if(N <= 3 || A == 2){
                while(M > N / 2){
                    int u = rng.nextInt(N);
                    int v = rng.nextInt(N);
                    par.set(u, par.get(u) ^ 1); par.set(v, par.get(v) ^ 1);
                    if(u != v && par.get(u) == par.get(v))o += 4 * par.get(u) - 2;
                    edges.add(new _A4D_Solution.Pair<Integer, Integer>(u, v));
                    M--;
                }
                ArrayList<Integer> oo = new ArrayList<Integer>();
                for(int i = 0; i < N; i++)
                    if(par.get(i) == 1)oo.add(i);
                Collections.shuffle(oo);
                for(int i = 0; i < oo.size(); i += 2){
                    int u = oo.get(i);
                    int v = oo.get(i + 1);
                    par.set(u, par.get(u) ^ 1); par.set(v, par.get(v) ^ 1);
                    if(u != v && par.get(u) == par.get(v))o += 4 * par.get(u) - 2;
                    edges.add(new _A4D_Solution.Pair<Integer, Integer>(u, v));
                    M--;
                }
            }else{
                while(M > A){
                    int u = rng.nextInt(N);
                    int v = rng.nextInt(N);
                    par.set(u, par.get(u) ^ 1); par.set(v, par.get(v) ^ 1);
                    if(u != v && par.get(u) == par.get(v))o += 4 * par.get(u) - 2;
                    edges.add(new _A4D_Solution.Pair<Integer, Integer>(u, v));
                    M--;
                }
                if(A == 1 && o <= 2){
                    ArrayList<Integer> ee = new ArrayList<>();
                    for(int i = 0; i < N; i++){
                        if(par.get(i) == 0)ee.add(i);
                    }
                    int u = ee.get(0);
                    int v = ee.get(1);
                    par.set(u, par.get(u) ^ 1); par.set(v, par.get(v) ^ 1);
                    if(u != v && par.get(u) == par.get(v))o += 4 * par.get(u) - 2;
                    edges.add(new _A4D_Solution.Pair<Integer, Integer>(u, v));
                    M--;
                }
            }
            checkSoln();
        }

    }

    public static void limitTime() {
        try {
            for (long recall = -1; elapsed < 1000000000L * timeLimit && recall != elapsed;) {
                recall = elapsed;
                Thread.sleep(1000 * timeLimit << 1);
            }
            exitCase("TimeLimitExceeded");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void main(String args[]) throws Exception {
        if (args.length == 1)
            rng.setSeed(batch = Integer.parseInt(args[0]));
        if (1 <= batch && batch <= nBatch) {
            System.out.println("Running on Batch " + batch);
            new Thread(() -> limitTime()).start();
            processBatch();
            System.out.printf("Accepted on Batch " + batch + " in %.9fs\n", elapsed * 1e-9);
            System.exit(0);
        }
        BufferedReader fin = new BufferedReader(new FileReader("EnterIDandLanguage.txt"));
        String ID = fin.readLine().split("\\s+", 2)[0];
        fin.close();
        if (new File("_A4D_" + ID + ".java").exists()) {
            fin = new BufferedReader(new FileReader("_A4D_" + ID + ".java"));
            best = Integer.parseInt(fin.readLine().split("\\s+", 4)[2]);
            fin.close();
        }
        new FileOutputStream("WrongAnswer.txt").close();
        new FileOutputStream("TimeLimitExceeded.txt").close();
        for (batch = 1; batch <= nBatch; total += weight[batch], ++batch) {
            Process process = new ProcessBuilder("java", "_A4D", "" + batch).start();
            fin = new BufferedReader(new InputStreamReader(process.getInputStream()));
            fin.lines().forEach(System.out::println);
            fin.close();
            fin = new BufferedReader(new InputStreamReader(process.getErrorStream()));
            fin.lines().forEach(System.out::println);
            fin.close();
            if (process.waitFor() == 0)
                score += weight[batch];
        }
        if (best <= score) {
            BufferedWriter fout = new BufferedWriter(new FileWriter("_A4D_" + ID + ".java"));
            fout.write("// " + ID + " " + score + "\n");
            fout.close();
            new ProcessBuilder("cmd", "/c", "echo // %COMPUTERNAME% %USERNAME%>>_A4D_" + ID + ".java").start()
                    .waitFor();
            fout = new BufferedWriter(new FileWriter("_A4D_" + ID + ".java", true));
            fin = new BufferedReader(new FileReader("_A4D_Solution.java"));
            fout.write(fin.lines().collect(java.util.stream.Collectors.joining(System.lineSeparator())));
            fin.close();
            fout.close();
        }
        System.out.print("Tentative score = " + (score / (double) total) + "/1\n");
        System.exit(0);
    }
}
